File: Algoritmos/ListaEnlazada.cpp
#include <iostream>
using namespace std;
struct Nodo
{
    int dato;
    Nodo *siguiente;
    Nodo(int valor)
    {
        dato = valor;
        siguiente = nullptr;
    }
};
class ListaEnlazada
{
private:
    Nodo *cabeza;

public:
    ListaEnlazada()
    {
        cabeza = nullptr;
    }
    void insertarInicio(int valor)
    {
        Nodo *nuevo = new Nodo(valor);
        nuevo->siguiente = cabeza;
        cabeza = nuevo;
        cout << "Insertado al inicio: " << valor << endl;
    }
    void insertarFinal(int valor)
    {
        Nodo *nuevo = new Nodo(valor);
        if (cabeza == nullptr)
        {
            cabeza = nuevo;
        }
        else
        {
            Nodo *temp = cabeza;
            while (temp->siguiente != nullptr)
            {
                temp = temp->siguiente;
            }
            temp->siguiente = nuevo;
        }
        cout << "Insertado al final: " << valor << endl;
    }
    void eliminarInicio()
    {
        if (cabeza == nullptr)
        {
            cout << "Error: Lista vacia" << endl;
            return;
        }
        Nodo *temp = cabeza;
        cabeza = cabeza->siguiente;
        cout << "Eliminado del inicio: " << temp->dato << endl;
        delete temp;
    }
    void mostrar()
    {
        if (cabeza == nullptr)
        {
            cout << "Lista vacia" << endl;
            return;
        }
        cout << "Lista final:" << endl;
        Nodo *temp = cabeza;
        while (temp != nullptr)
        {
            cout << temp->dato << endl;
            temp = temp->siguiente;
        }
    }
    ~ListaEnlazada()
    {
        while (cabeza != nullptr)
        {
            Nodo *temp = cabeza;
            cabeza = cabeza->siguiente;
            delete temp;
        }
    }
};
int main()
{
    ListaEnlazada lista;
    cout << "=== Insertando al inicio ===" << endl;
    lista.insertarInicio(8);
    lista.insertarInicio(4);
    cout << "\n=== Insertando al final ===" << endl;
    lista.insertarFinal(11);
    cout << "\n=== Eliminando primer nodo ===" << endl;
    lista.eliminarInicio();
    cout << "\n=== Estado final ===" << endl;
    lista.mostrar();
    return 0;
}

File: Algoritmos/ListaEnlazada.py
class Nodo:
    def __init__(self, valor):
        self.dato = valor
        self.siguiente = None

class ListaEnlazada:
    def __init__(self):
        self.cabeza = None

    def insertar_inicio(self, valor):
        nuevo = Nodo(valor)
        nuevo.siguiente = self.cabeza
        self.cabeza = nuevo
        print(f"Insertado al inicio: {valor}")

    def insertar_final(self, valor):
        nuevo = Nodo(valor)
        if self.cabeza is None:
            self.cabeza = nuevo
        else:
            temp = self.cabeza
            while temp.siguiente:
                temp = temp.siguiente
            temp.siguiente = nuevo
        print(f"Insertado al final: {valor}")

    def eliminar_inicio(self):
        if self.cabeza is None:
            print("Error: Lista vacia")
            return
        eliminado = self.cabeza.dato
        self.cabeza = self.cabeza.siguiente
        print(f"Eliminado del inicio: {eliminado}")

    def mostrar(self):
        if self.cabeza is None:
            print("Lista vacia")
            return
        print("Lista final:")
        temp = self.cabeza
        while temp:
            print(temp.dato)
            temp = temp.siguiente

if __name__ == "__main__":
    lista = ListaEnlazada()
    print("=== Insertando al inicio ===")
    lista.insertar_inicio(8)
    lista.insertar_inicio(4)
    print("\n=== Insertando al final ===")
    lista.insertar_final(11)
    print("\n=== Eliminando primer nodo ===")
    lista.eliminar_inicio()
    print("\n=== Estado final ===")
    lista.mostrar()